src/service/video-manager-impl.js

```diff
@@ -436,6 +540,136 @@ class VideoEntry {
+      vidStyle.width = SCALE * this.initialWidth_ + 'px';
+      vidStyle.borderRadius = '6px';
+      vidStyle.left = '20px';
+      vidStyle.boxShadow = '0px 0px 20px 0px rgba(0, 0, 0, 0.3)';
+      if (this.minimizePosition_ == MinimizePositions.BOTTOM) {
+        this.minimizePosition_ = MinimizePositions.TOP;
+        vidStyle.bottom = '20px';
+        vidStyle.top = 'auto';
+      } else {
+        this.minimizePosition_ = MinimizePositions.BOTTOM;
+        vidStyle.top = '20px';
+        vidStyle.bottom = 'auto';
+      }
+      return;
+    }
+
+    const docViewTop = newPos.viewportRect.top;
+    const docViewBottom = newPos.viewportRect.bottom;
+
+    const elemTop = newPos.positionRect.top;
+    const elemBottom = newPos.positionRect.bottom;
+
+    let inViewportHeight = 0;
+
+    // Calculate height currently displayed
+    if (elemTop <= docViewTop) {
+      inViewportHeight = elemBottom - docViewTop;
+      this.minimizePosition_ = MinimizePositions.TOP;
+    } else if (elemBottom >= docViewBottom) {
+      inViewportHeight = docViewBottom - elemTop;
+      this.minimizePosition_ = MinimizePositions.BOTTOM;
+    } else {
+      this.minimizePosition_ = MinimizePositions.INVIEW;
+    }
+
+    if (this.minimizePosition_ != MinimizePositions.INVIEW
+      && this.getPlayingState() == PlayingStates.PLAYING_MANUAL
+      && this.minimizePosition_ != MinimizePositions.DEFAULT) {
+
+      // Minimize the video
+
+      this.video.hideControls();
+
+      vidElement.classList.add('i-amphtml-dockable-video-minimizing');
+
+      vidStyle.height = mapRange(inViewportHeight,
+          0, this.initialHeight_,
+          SCALE * this.initialHeight_, this.initialHeight_
+      ) + 'px';
+
+      vidStyle.width = mapRange(inViewportHeight,
+          0, this.initialHeight_,
+          SCALE * this.initialWidth_, this.initialWidth_
+      ) + 'px';
+
+      vidStyle.borderRadius = mapRange(inViewportHeight,
+          this.initialHeight_, 0,
+          0, 6
+      ) + 'px';
+
+      const shdw = mapRange(inViewportHeight,
+          this.initialHeight_, 0,
+          0, 0.3
+      );
+
+      vidStyle.boxShadow = '0px 0px 20px 0px rgba(0, 0, 0, ' + shdw + ')';
+
+      vidStyle.left = mapRange(inViewportHeight,
+          this.initialHeight_, 0,
+          this.initialLeft_, 20
+      ) + 'px';
+
+      // Different behavior based on whether the video got minimized
+      // from the top or the bottom
+      if (this.minimizePosition_ == MinimizePositions.TOP) {
+
+        vidStyle.top = mapRange(inViewportHeight,
+            this.initialHeight_, 0,
+            0, 20
+        ) + 'px';
+
+        vidStyle.bottom = 'auto';
+
+      } else {
+
+        vidStyle.bottom = mapRange(inViewportHeight,
+            this.initialHeight_, 0,
+            0, 20
+        ) + 'px';
+
+        vidStyle.top = 'auto';
+      }
+
+      // TODO(@wassim) Make minimized video draggable
+
+    } else if (this.minimizePosition_ == MinimizePositions.INVIEW) {
```

Comment:
can we split into three methods:
1. startDocking_
2. duringDocking_
3. endDocking_ 

and have this function just do state checks and call the appropriate sub function?